src/NonBlocking/ConcurrentDictionary/DictionaryImpl`3.cs

```diff
@@ -1316,10 +1318,9 @@ private static bool CopySlot(ref Entry oldEntry, DictionaryImpl<TKey, TKeyStore,
             // Align up to a power of 2
             newsz = Util.AlignToPowerOfTwo(newsz);
 
-            // Size calculation: 2 words (K+V) per table entry, plus a handful.  We
-            // guess at 32-bit pointers; 64-bit pointers screws up the size calc by
-            // 2x but does not screw up the heuristic very much.
-            int kBs4 = (((newsz << 1) + 4) << 3/*word to bytes*/) >> 12/*kBs4*/;
```

Comment:
The bug was here. Left shifts could turn the large size into 0, leading to allocating multiple candidates (since we assumed that is chap and fast).

The rest is just rationalizing and cleaning up this logic a bit.

Reply:
This mostly matters just in corner cases when dealing with abnormally huge dictionaries.